### PR TITLE
fix(auth): add inflightPromise in token orchestrator

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

The main issue is due to a race condition between `attemptCompleteOAuthFlow` and `TokenOrchestrator.getTokens`

Ideally when there is an inflight OAuth flow, `attemptCompleteOAuthFlow` should always be called first to register any upcoming async calls coming from `TokenOrchestrator.getTokens`. This behavior is consistent in React but the affected frameworks are bundled different.

The fix at the moment is to check if `OAuth` is inflight directly from the TokenOrchestrator and add an inflight promise if so.

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts [hunk 1]
```diff
@@ -23,13 +23,23 @@ import {
...
 	tokenStore?: AuthTokenStore;
 	tokenRefresher?: TokenRefresher;
 	waitForInflightOAuth: () => Promise<void> = async () => {
 		// no-op
+		if (!(await oAuthStore.loadOAuthInFlight())) {
+			return;
+		}
+		
+		return new Promise<void>((resolve, _reject) => {
+			addInflightPromise(resolve);
```
Comment (HuiSF) [inline packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts:40]: Should we remove the `addInflightPromise()` call from `attempCompleteOAuthFlow` then?

File: packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts [hunk 2]
```diff
@@ -23,13 +23,23 @@ import {
...
 } from './types';
+import { oAuthStore } from '../utils/oauth';
+import { addInflightPromise } from '../utils/oauth/inflightPromise';
 
 export class TokenOrchestrator implements AuthTokenOrchestrator {
 	private authConfig?: AuthConfig;
 	tokenStore?: AuthTokenStore;
 	tokenRefresher?: TokenRefresher;
 	waitForInflightOAuth: () => Promise<void> = async () => {
 		// no-op
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts:34]: Nit: Not a no-op anymore

**Review by jimblanc — COMMENTED** [review]
If we're planning on hot-fixing this, should target `release` branch

File: packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts (hunk 1 above)
Comment (israx) [reply packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts:40]: yeah good point

Comment (HuiSF) [reply packages/auth/src/providers/cognito/tokenProvider/TokenOrchestrator.ts:40]: `cognitoUserPoolsTokenProvider.setWaitForInflightOAuth` should be removed as well as its single use case is being replace by this. 👍

**Pull request closed without merging**